src/pocketmine/network/upnp/UPnP.php

```diff
@@ -19,63 +19,183 @@
 namespace pocketmine\network\upnp;
 
 use pocketmine\utils\Internet;
-use pocketmine\utils\Utils;
-use function class_exists;
-use function is_object;
+use SimpleXMLElement;
+use function socket_create;
+use function socket_set_option;
+use function socket_recvfrom;
+use function socket_sendto;
+use function trim;
+use function strlen;
+use function preg_match;
+use function parse_url;
+use const AF_INET;
+use const SOCK_DGRAM;
+use const SOL_UDP;
+use const IPPROTO_IP;
+use const IP_MULTICAST_IF;
+use const IP_MULTICAST_LOOP;
+use const IP_MULTICAST_TTL;
+use const SOL_SOCKET;
+use const SO_RCVTIMEO;
 
 abstract class UPnP{
+	/** @var string|null */
+	private static $serviceURL;
+	
+	public static function getControlUrl() : string{
+		try{
+			$socket = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+			socket_set_option($socket, SOL_SOCKET, SO_RCVTIMEO, ["sec" => 3, "usec" => 0]);
+			$contents =
+				"M-SEARCH * HTTP/1.1\r\n" .
+				"MX: 2\r\n" .
+				"HOST: 239.255.255.250:1900\r\n" .
+				"MAN: \"ssdp:discover\"\r\n" .
+				"ST: upnp:rootdevice\r\n\r\n";
+			socket_sendto($socket, $contents, strlen($contents), 0, "239.255.255.250", 1900);
+			socket_recvfrom($socket, $buffer, 1024, 0, $responseHost, $responsePort);
+			socket_close($socket);
+		}catch(\Exception $e){
+			throw new \RuntimeException("Socket error: " . $e->getMessage());
+		}
+		if($buffer === false || !preg_match('/location\s*:\s*(.+)\n/i', $buffer, $matches)){
+			throw new \RuntimeException("Unable to find the router. Ensure that network discovery is enabled in Control Panel.");
+		}
+		$location = trim($matches[1]);
+		$url = parse_url($location);
+		if($url === false){
+			throw new \RuntimeException("Failed to parse the router's url: {$location}");
+		}
+		$response = Internet::getURL($location, 3, [], $err, $headers, $httpCode);
+		if($response === false){
+			throw new \RuntimeException("Unable to access XML: {$err}");
+		}
+		if($httpCode !== 200){
+			throw new \RuntimeException("Unable to access XML: {$response}");
+		}
+
+		$defaultInternalError = libxml_use_internal_errors(true);
+		try{
+			$root = new \SimpleXMLElement($response);
+		}catch(\Exception $e){
+			throw new \RuntimeException("Broken XML.");
+		}
+		libxml_use_internal_errors($defaultInternalError);
+		
+		$xpathResult = $root->xpath(
+			'//device[deviceType="urn:schemas-upnp-org:device:InternetGatewayDevice:1"]' .
+			'/deviceList/device[deviceType="urn:schemas-upnp-org:device:WANDevice:1"]' .
+			'/deviceList/device[deviceType="urn:schemas-upnp-org:device:WANConnectionDevice:1"]' .
+			'/serviceList/service[serviceType="urn:schemas-upnp-org:service:WANIPConnection:1"]' .
+			'/controlURL'
+		);
+		if($xpathResult === false || count($xpathResult) === 0){
+			// if result is false or [], i.e. error or notfound
+			throw new \RuntimeException("Your router does not support portforwarding");
+		}
+		$controlURL = (string)($xpathResult[0]);
+		$serviceURL = "{$url['host']}:{$url['port']}/{$controlURL}";
```

Comment:
I'd prefer if this used concat operators or even `sprintf()`, this is difficult to read